broadinstitute/single_cell_portal_core
# Adding user bookmarks (SCP-5431)

#### BACKGROUND & CHANGES
This update adds the ability for users to bookmark selected views in the Explore tab.  This applies to all Explore-based visualizations that persist state into the URL bar (e.g. scatter, violin, dot plot, differential expression).  Along with a name, users can provide a description to annotate their bookmarks, which can also be updated and deleted from the same form.  Clicking "Save" without prompting a name will pre-fill the current URL path + query string as the name.  Additionally, users can see all their bookmarks in a list and go directly to each visualization - whether in the same study or a different one.

Here's a 1 min video showing the basic functionality (note - this only covers normal scatter plots, but all Explore tab visualizations are supported):

https://github.com/broadinstitute/single_cell_portal_core/assets/729968/8b968c9c-1b03-43fe-bcfb-54e131333108

#### MANUAL TESTING
1. Boot as normal and load any study with visualizations
2. Click the bookmark (star) icon and confirm you are prompted to sign in
3. After signing in, click the bookmark icon again and then save without providing any values
4. Confirm the save succeeds, then click "See bookmarks" to pull up the list, confirming that the new bookmark is listed with the path as the name
5. Close the list and then provide a name/description for this bookmark and save, confirming it succeeds
6. Load different visualization types (violin, dotplot, etc.) and confirm you can bookmark/load then without issue
7. Delete any bookmark, then re-open the bookmarks list and confirm it no longer appears
